### PR TITLE
Apply console context to deprecate warnings

### LINKED ISSUE #9: TypeError: Illegal invocation when using Metal.deprecate
[Example](http://requirebin.com/?gist=500332fcbe0ed40b7268)

It's because `console` methods need to have `console` as the context, and calling `deprecate._warn` doesn't do that.

### PR DESCRIPTION
**Pull request by paulfalgout**
To ensure console compatibility

Resolves #9

Solution from https://github.com/marionettejs/backbone.marionette/pull/2459

### CODE REVIEW

**Comment by megawac** [comment]
Looks good

**Comment by jamiebuilds** [comment]
Is there much of a point in exposing `_console` ?

**Comment by paulfalgout** [comment]
@thejameskyle hmm I don't suppose so.  I mainly just copy-pasta'd this from the `Mn` commit.. though I suppose it'd be hard to do this without the exposure:  https://github.com/paulfalgout/backbone-metal/commit/c0a302150493f1889237db8e547bd3ccb4907927#diff-8cb4b8007541a097153799972ea833d9R4

But that essentially covered already by the additional test below.. _I think_  

Seems like it might as well be local to `_warn`?  @cmaher @megawac thoughts on the `_console` exposure question?

**Comment by jamiebuilds** [comment]
Why not just this?

``` js
describe('deprecate', function() {
  beforeEach(function() {
    spy(Metal.deprecate, '_warn');
    Metal.deprecate._cache = {};
  });

  describe('Metal.deprecate._warn', function() {
    beforeEach(function() {
      Metal.deprecate._warn('foo');
    });

    it('should `console.warn` the message', function() {
      expect(Metal.deprecate._warn)
        .to.have.been.calledOnce
        .and.calledOn(console)
        .and.calledWith('foo');
    });
    // ...
  });
  // ...
});
```

**Comment by cmaher** [comment]
Pretty sure `_warn` is being called on `Marionette.deprecate`, not `console`, hence the indirection.

**Comment by jamiebuilds** [comment]
huh? 

``` js
return warn.apply(deprecate._console, arguments);
```

**Comment by cmaher** [comment]
mm, I thought you meant not exposing `_console` at all.

**Comment by jamiebuilds** [comment]
Yeah, I do mean that. Just store it as a local variable. I was quoting the code in this PR.

``` diff
- deprecate._console = typeof console !== 'undefined' ? console : {};
+ var _console = typeof console !== 'undefined' ? console : {};

  deprecate._warn = function() {
-   var warn = deprecate._console.warn || deprecate._console.log || _.noop;
-   return warn.apply(deprecate._console, arguments);
+   var warn = _console.warn || _console.log || _.noop;
+   return warn.apply(_console, arguments);
  };
```

**Comment by paulfalgout** [comment]
Sorry for delay on this.  I gave the suggested a try, but haven't had the chance to look at the failing test.

Also I moved the local `_console` into the function..  @thejameskyle think it is better to declare it outside of the `_warn`?

**Comment by paulfalgout** [comment]
Ok so defining _console inside warn made testing _.noop impossible.

Also not attaching _console... I can't get `.and.calledOn(console)`  I'm not sure if that test is really necessary?  Assuming not, this might be ready to go.

**Comment by jasonLaster** [comment]
what's left here?

**Comment by paulfalgout** [comment]
If the tests here are acceptable this is ready

**Comment by jasonLaster** [comment]
:+1:

File: src/backbone-metal.js [hunk 1]
```diff
@@ -451,7 +460,10 @@ deprecate._format = function(prev, next) {
  * @memberOf deprecate
  * @param {*...} - The values to warn in the console.
  */
-deprecate._warn = typeof console !== 'undefined' && (console.warn || console.log) || _.noop;
+deprecate._warn = function() {
+  var warn = _console.warn || _console.log || _.noop;
```
Comment (jamiebuilds) [inline src/backbone-metal.js]: Couldn't you pull this part out?

``` js
var _console = typeof console !== 'undefined' ? console : {};
var _warn = _console.warn || _console.log || _.noop;

deprecate._warn = function() {
  return _warn.apply(_console, arguments);
};
```

Comment (paulfalgout) [inline src/backbone-metal.js]: Well so should it really just be:

``` js
var _console = typeof console !== 'undefined' ? console : {};
deprecate._warn = _console.warn || _console.log || _.noop;
```

or

``` js
var _console = typeof console !== 'undefined' ? console : {};
var _warn = _console.warn || _console.log || _.noop;

deprecate._warn = _warn;
```

Or am I missing something?  thoughts?

Comment (yethee) [inline src/backbone-metal.js]: Chrome throws error:

```
TypeError: Illegal invocation
```

`this` should be reference to console object.

Comment (paulfalgout) [inline src/backbone-metal.js]: ah so

``` js
deprecate._warn = _.bind(_warn, _console);
```

Comment (jamiebuilds) [inline src/backbone-metal.js]: that wouldn't be called with the `console` context.

**Comment by jamiebuilds** [comment]
:+1:

**Pull request merged**